### ISSUE TITLE
pytest fixtures with scope="function" should be executed each time rerun is called    

### ISSUE THREAD

**Issue by amit-selector** [pytest-dev/pytest-rerunfailures#198]
Based on 10.2 version, it is noticed that when a failed testcase is rerun, the fixtures with "function" scope are not executed each rerun. In my understanding is such fixtures should be called first each time a testcase is rerun. It appears they are not getting run since I have a counter that is initialized to zero in such a fixture and the counter continues to tick up from where it was in the failed test's previous run.

**Comment by icemac** [comment]
Thank you for writing this issue. It seems strange that this might have been unnoticed for such a long time.
Could you please come up with a minimal example showing your observations?

**Comment by amit-selector** [comment]
You are right - I did a basic test and it seems to work fine. My
requirement is a bit more involved so let me investigate and get back to
you if I can get a simplified example (or perhaps uncover some problem in
my code). Thanks for getting in touch quickly!

**Comment by dcermak** [comment]
I have run into a very similar issue, however the issue was not that the fixtures would not run. It was more that pytest-rerunfailures would only partially reset the data that were used to parametrize my tests. I have a dataclass that contains a string and a TemporaryDirectory which would be both set in my fixture. Rerunfailures would then restart my test with an instance of that class where the string contained the previously set value, but the temporary directory was not copied over. It caused some very weird failures.

I have tried to create a minimal reproducer for this, but so far the only that I have managed is the following:
https://gist.github.com/dcermak/8b7a3118fe60659d4e38f825a2f26de3

It demonstrates some odd inconsistencies with module level fixtures: rerunfailures will sometimes pick the same object to restart the test or sometimes rerun the fixture, although this is a module level fixture.

Unfortunately I don't know if this is a different issue or just a red herring.
